### ISSUE TITLE
Error: dim(X) must have positive length

### ISSUE THREAD

**Issue by ghost** [statnmap/HatchedPolygons#3]
Hi Sébastien!

I just ran into an error that only occurs when a map I want to make includes a specific polygon. This is the error when I run hatched.SpatialPolygons:

Error in apply(lines.hatch, 1, function(x) Line(cbind(c(x[1], x[3]), c(x[2],  : 
  dim(X) must have a positive length

I found that someone else has run into this same error when using hatched.SpatialPolygons. Here is the link where it is mentioned by 'mcSamuelDataSci': https://github.com/mtennekes/tmap/issues/49 

If you want I can send you the shapefile and data, in order to reproduce the error. 

Anyway, what's strange is that in the same dataset, whenever I plot polygons of other areas, it works, but for whatever reason one of the areas, when included, throws that error. Leaflet itself is able to plot the polygon in question. Only hatched.SpatialPolygons throws the error. 

Thanks again for all your help in the past!

**Comment by statnmap** [comment]
Hi,
If you can extract the polygon that leads to this problem, that would be easier to try to debug.
Thank you

**Comment by Sboyer28** [comment]
Hi Sebastian, 

I received the same error as jcutler79 above. 

I am working with a SpatialPolygonsDataFrame, which is a census bureau tigerline shapefile I downloaded from their site:
> class(county_geodata)
[1] "SpatialPolygonsDataFrame"
attr(,"package")
[1] "sp"

I am trying to use hatched.SpatialPolygons in leaflet, but when I actually hatch the polygons I subset, I receive the error:

>hatch_vacc_line <- hatch_vacc %>%
>    subset(Series_complete < 33)
>vacc.hatch <- hatched.SpatialPolygons(hvl.sf, density = c(6,4), angle = c(45,135))

"Error in apply(lines.hatch, 1, function(x) Line(cbind(c(x[1], x[3]), c(x[2], :
dim(X) must have a positive length"

Thanks in advance for any help you can give.

**Comment by Sboyer28** [comment]
edit: wrong (x) in the hatched.SpatialPolygons function:
>hatch_vacc_line <- hatch_vacc %>%
 >   subset(Series_complete < 33)
>vacc.hatch <- hatched.SpatialPolygons(hatch_vacc_line, density = c(6,4), angle = c(45,135))

**Comment by statnmap** [comment]
Hi, 
Could you please filter the spatial dataset to isolate the polygon giving problems and share it there ?
So that I can reproduce the error.
Maybe you can download the dataset using {tigris} and then filter: https://cran.r-project.org/web/packages/tigris/index.html
Thank you.

**Comment by Sboyer28** [comment]
Hi Sebastian,

Thank you for your reply.
So I am not sure how to go about filtering the polygon or polygons that are causing the issues. I am dealing with a couple thousand polygons.

Thanks.

**Comment by statnmap** [comment]
Hi,
You can subset by finding which ones are problematic with a loop or `lapply()`. But even if you do not subset to the polygons having problems, I need a way to reproduce it.
Can you provide the smallest example code you can, with associated dataset so that I can have the same error than you on my computer, without changing this code?
Thanks.

**Comment by statnmap** [comment]
Hi,
I dealt with a similar case today.
The difficulty with multi polygons is that some of them are so small that the density chosen is surely not big enough. 
Unfortunately, there is presently no way to calculate in advance the density required by each polygon.
I used a big value of `density` and it worked.
If you can separate polygons with `st_cast()`, then you can test which value better suits your needs for each of the polygons.
But it will be a little long to explore that by hand.
Otherwise, you can simplify your geometries, even remove too small polygons to avoid such difficulty.
You can try some parametrization of `keep` in `rmapshaper::ms_simplify()`.

Some way of dealing with this.

```r
# Separate polygons
sf_layer_polygon <- sf::st_cast(sf_layer, "POLYGON") %>%
  mutate(name = as.character(1:n()))

ggplot(sf_layer_polygon) +
  geom_sf(aes(fill = name))

# Try to find out which layer has a problem
for (i in 1:nrow(sf_layer_polygon)) {
    print(i)

    layer <- sf_layer_polygon %>% 
      slice(i)

   layer.sp <- as(layer, "Spatial")

    # Try different densities, the smallest the polygon, the biggest the density
   hatched <- hatched.SpatialPolygons(layer, density = c(10000), angle = c(135))

  plot(layer["ID"])
  # plot(layer["ID"], type = "p")
  # points(layer.sp@polygons[[1]]@Polygons[[1]]@coords)


}
```
I opened a new issue to try to see if we can propose `density` value to avoid this problem: https://github.com/statnmap/HatchedPolygons/issues/6

**Comment by warnes** [comment]
This error is resolved by patch #9.  It now generates this warning when a polygon is too small:
```
In polygonRingHolesLines(polys[[j]], density = density[j],  ... :
  Polygon too small to contain any lines.  Consider increasing 'density'.
```
